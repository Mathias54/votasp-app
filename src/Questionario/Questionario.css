.questionario__container {
  display: flex;
  flex-direction: column;
  height: calc(100% - 63px);
  padding: 20px;
}

.questionario__actions-container {
  display: flex;
  justify-content: space-around;
  margin-top: auto;
}
