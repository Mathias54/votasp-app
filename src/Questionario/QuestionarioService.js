import firebase from 'firebase/app';

export const saveAnswer = (answer) => {
  const userId = firebase.auth().currentUser.uid;
  return firebase
    .firestore()
    .collection('answers')
    .doc(userId)
    .set(answer, { merge: true });
}

export const watchAnswers = () => {
  const userId = firebase.auth().currentUser.uid;
  return firebase
    .firestore()
    .collection('answers')
    .doc(userId);
}

