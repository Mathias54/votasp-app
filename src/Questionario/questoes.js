const questoes = [
  {
    "id": 1,
    "question": "Legalização do aborto para todos os casos."
  },
  {
    "id": 2,
    "question": "Adoção de crianças por casais do mesmo sexo."
  },
  {
    "id": 3,
    "question": "Debate de gênero nas escolas."
  },
  {
    "id": 4,
    "question": "Liberação dos portes de armas no Brasil."
  },
  {
    "id": 5,
    "question": "Adoção de cotas raciais nas universidades públicas."
  },
  {
    "id": 6,
    "question": "Privatização de serviços públicos como educação, saúde e cultura."
  },
  {
    "id": 7,
    "question": "A reforma trabalhista de 2017 modernizou as relações de trabalho do Brasil."
  },
  {
    "id": 8,
    "question": "Desapropriação de prédios ociosos para moradia popular."
  },
  {
    "id": 9,
    "question": "O Brasil é um país racista."
  },
  {
    "id": 10,
    "question": "Criminalização da homofobia no Brasil."
  },
  {
    "id": 11,
    "question": "Redução da maioridade penal para 16 anos."
  },
  {
    "id": 12,
    "question": "Definição exclusiva de família como homem mulher e filhos"
  },
  {
    "id": 13,
    "question": "Campanhas eleitorais financiadas unicamente com dinheiro público."
  },
  {
    "id": 14,
    "question": "Leis para garantir a igualdade de salários entre homens e mulheres"
  },
  {
    "id": 15,
    "question": "Intervenção militar para combater a criminalidade."
  },
  {
    "id": 16,
    "question": "Regulamentação dos meios de comunicação."
  },
  {
    "id": 17,
    "question": "Os ricos deveriam pagar mais impostos."
  },
  {
    "id": 18,
    "question": "Privatização da previdência social."
  },
  {
    "id": 19,
    "question": "Privatização completa da Petrobrás"
  },
  {
    "id": 20,
    "question": "Em briga de marido e mulher ninguém mete a colher."
  },
  {
    "id": 21,
    "question": "O financimento empresarial de campanhas políticas deixa o candidato com rabo preso."
  },
  {
    "id": 22,
    "question": "Fim do foro privilegiado."
  },
  {
    "id": 23,
    "question": "Legalização do consumo da maconha."
  },
  {
    "id": 24,
    "question": "Garantia pelo governo de aposentadoria para todas e todos os brasileiros."
  },
  {
    "id": 25,
    "question": "Saúde pública de qualidade e gratuita para toda a população."
  },
  {
    "id": 26,
    "question": "Isenção de impostos de instituições religiosas."
  },
  {
    "id": 27,
    "question": "Garantia de mesmos direitos dos brasileiros(as) para imigrantes."
  },
  {
    "id": 28,
    "question": "Reforma agrária."
  },
  {
    "id": 29,
    "question": "A conciliação é uma medida útil para ajudar nos casos de violência doméstica e familiar"
  },
  {
    "id": 30,
    "question": "Construções de barragens para solucionar a crise hídrica."
  },
  {
    "id": 31,
    "question": "Fim do auxílio moradia para juízes."
  },
  {
    "id": 32,
    "question": "Maior carga tributária para empresas estrangeiras com o objetivo de beneficiar empresas brasileiras."
  },
  {
    "id": 33,
    "question": "Pena de morte."
  },
  {
    "id": 34,
    "question": "As Universidades Públicas deveriam ser paga por quem pode pagar."
  },
  {
    "id": 35,
    "question": "Construção de moradias populares na periferia mesmo antes de garantir atendimento básico como centros de saúde e escolas"
  },
  {
    "id": 36,
    "question": "Grandes empreendimentos podem iniciar suas obras mesmo antes de conseguirem permissão ambiental"
  },
  {
    "id": 37,
    "question": "Crescimento econômico é mais importante que igualdade social"
  },
  {
    "id": 38,
    "question": "Interferência do estado na economia."
  },
  {
    "id": 39,
    "question": "Redução da verba de gabinete para parlamentares"
  },
  {
    "id": 40,
    "question": "O governo tem um papel essencial na regulação dos mercados para garantir os direitos dos consumidores."
  }
]

export default questoes;
