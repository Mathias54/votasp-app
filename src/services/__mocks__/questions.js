export const getQuestions = jest.fn(() => Promise.resolve({
  data: {
    perguntas: [
      {
        id: 1,
        question: 'Legalização do aborto para todos os casos.'
      },
      {
        id: 2,
        question: 'Adoção de crianças por casais do mesmo sexo.'
      },
    ],
  },
}));
