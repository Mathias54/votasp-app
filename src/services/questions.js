import axios from 'axios';

export const getQuestions = () => axios.get('https://questions.api.votasp.org.br/');

