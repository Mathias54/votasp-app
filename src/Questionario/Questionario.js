import React, { Component } from 'react';
import classnames from 'classnames';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import PageLayout from '../components/PageLayout/PageLayout';
import Pergunta from '../components/Pergunta/Pergunta';
import ProgressBar from '../components/ProgressBar/ProgressBar';
import QuestionsMenu from '../components/QuestionsMenu/QuestionsMenu';
import questoes from './questoes'
import store from '../redux/store';
import { INDIFERENTE } from '../constants/respostas';
import { saveAnswer, watchAnswers, getCurrentUser } from './QuestionarioService';
import { storePerguntas } from '../redux/modules/perguntas';
import { storeQuestionario } from '../redux/modules/questionario';

import './Questionario.css';

export class RawQuestionario extends Component {

  state = {
    isAnswering: false,
    userAnswers: [],
    userCollection: undefined,
  };

  componentDidMount() {
    store.dispatch(storePerguntas(questoes));
    getCurrentUser().then(this.saveUser);
  }

  saveUser = (doc) => {
    const user = doc.data();
    const isCandidate = user.role === 'candidate';
    const userCollection = isCandidate ? 'candidate_answers' : 'voter_answers';

    watchAnswers(userCollection).onSnapshot(this.storeAnswers);

    this.setState({
      userCollection
    });
  }

  storeAnswers = (snapshot) => {
    const data = snapshot.data();
    const userAnswers = data ? Object.keys(data)
      .map(answerKey => ({
        id: answerKey,
        answer: data[answerKey]
      })) : [];

    this.setState({
      userAnswers,
    });
  }

  pularQuestao = () => {
    return this.saveAnswer(INDIFERENTE.value).then(() => {
      this.proximaQuestao();
    });
  };

  responderQuestao = (event) => {
    this.saveAnswer(event.target.value).then(() => {
      this.proximaQuestao();
    });
  };

  saveAnswer = (answerValue) => {
    const { userCollection } = this.state;
    const { questionario } = this.props;
    const questionId = questionario.currentQuestion + 1;
    const answer = {
      [questionId]: answerValue,
    };

    this.setState({
      isAnswering: true,
    });

    return saveAnswer(answer, userCollection);
  }

  proximaQuestao = () => {
    const { perguntas, questionario } = this.props;

    if (questionario.currentQuestion === perguntas.length - 1) {
      this.props.history.push('/calculando-ranking');
      return;
    }

    store.dispatch(storeQuestionario({
      currentQuestion: questionario.currentQuestion + 1
    }));

    this.setState({
      isAnswering: false,
    });
  };

  nivelDoProgresso = () => {
    const { userAnswers } = this.state;
    const { perguntas } = this.props;

    if (userAnswers.length === 0) {
      return [
        {
          value: 0,
        }
      ]
    }
  
    const porcentagemDeProgresso = (userAnswers.length / perguntas.length) * 100 ;
    const porcentagemMinima = porcentagemDeProgresso > 50 ? 50 : porcentagemDeProgresso;
    const porcentagemAcimaDoMinimo = porcentagemMinima < 50 ? 0 : porcentagemMinima;

    const nivelDoProgresso = [
      {
        value: porcentagemMinima,
        color: '#fbdaab',
      },
      {
        value: porcentagemAcimaDoMinimo,
        color: '#feb557',
      }
    ];

    return nivelDoProgresso;
  }

  render() {
    const { isAnswering, userAnswers } = this.state;
    const { perguntas, questionario } = this.props;
    const { currentQuestion } = questionario;
    const [currentAnswer] = userAnswers.filter(answer => answer.id == currentQuestion + 1);
    const nivelDoProgresso = this.nivelDoProgresso();

    return (
      <PageLayout>
        <ProgressBar values={nivelDoProgresso}/>
        <QuestionsMenu
          userAnswers={userAnswers}
          questionsArray={questoes}
          currentQuestion={currentQuestion}
        />
        <div className={classnames(
          'questionario__container',
          { 'questionario__container--loading': isAnswering }
        )}>
          {perguntas.length &&
            <Pergunta
              pergunta={perguntas[currentQuestion]}
              responderQuestao={this.responderQuestao}
              isAnswering={isAnswering}
              userAnswer={currentAnswer ? currentAnswer.answer : undefined}
            />
          }

          <div className="questionario__actions-container">
            <button
              onClick={this.pularQuestao}
              className="btn btn-light"
              disabled={isAnswering}
            >
              Pular
            </button>
            <Link
              to="/calculando-ranking"
              className="btn btn-light"
              disabled={isAnswering || userAnswers.length < 20}
            >
              Calcular afinidade
            </Link>
          </div>
        </div>
      </PageLayout>
    );
  }
}

const mapStateToProps = ({ perguntas, questionario }) => ({
  perguntas,
  questionario,
});

export default connect(mapStateToProps)(RawQuestionario);
