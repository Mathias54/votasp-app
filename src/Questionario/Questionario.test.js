import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { getQuestions } from '../services/questions';

import { RawQuestionario } from './Questionario';

jest.mock('../services/questions');

const defaultProps = {
  perguntas: [
    {
      id: 1,
      question: 'Legalização do aborto para todos os casos.'
    },
    {
      id: 2,
      question: 'Adoção de crianças por casais do mesmo sexo.'
    },
  ],
}

describe('<Questionario />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const questionario = shallow(<RawQuestionario {...defaultProps} />);

    expect(toJson(questionario)).toMatchSnapshot();
  });

  it('should increment currentQuestion state', () => {
    const questionario = shallow(<RawQuestionario {...defaultProps} />);

    questionario.instance().pularQuestao();
    questionario.update();

    expect(questionario.state('currentQuestion')).toBe(1);
  });

  it('should not increment currentQuestion state when it reaches final question', () => {
    const questionario = shallow(<RawQuestionario {...defaultProps} />);

    const currentQuestion = defaultProps.perguntas.length - 1;
    questionario.setState({ currentQuestion });
    questionario.instance().pularQuestao();

    expect(questionario.state('currentQuestion')).toBe(1);
  });

  it('should call questions service to get all the questions', () => {
    const questionario = shallow(<RawQuestionario {...defaultProps} />);
    questionario.update();
    expect(getQuestions).toHaveBeenCalledTimes(1);
  });
});
