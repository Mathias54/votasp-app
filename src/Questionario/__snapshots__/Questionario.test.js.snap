// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Questionario /> should match snapshot 1`] = `
<PageLayout>
  <div
    className="questionario__container"
  >
    <Pergunta
      pergunta={
        Object {
          "id": 1,
          "question": "Legalização do aborto para todos os casos.",
        }
      }
    />
    <div
      className="questionario__actions-container"
    >
      <button
        className="btn btn-light"
        onClick={[Function]}
      >
        Pular
      </button>
      <button
        className="btn btn-light"
        disabled={true}
        onClick={[Function]}
      >
        Calcular afinidade
      </button>
    </div>
  </div>
</PageLayout>
`;
